packages/block-editor/src/components/use-on-block-drop/index.js

```diff
@@ -239,28 +232,92 @@ export default function useOnBlockDrop( targetRootClientId, targetBlockIndex ) {
 		clearSelectedBlock,
 	} = useDispatch( blockEditorStore );
 
+	const [ dropEventData, setDropEventData ] = useState( null );
+	const [ droppedFiles, setDroppedFiles ] = useState( null );
+	const [ droppedHTML, setDroppedHTML ] = useState( null );
+
+	useEffect( () => {
+		if ( dropEventData ) {
+			handleBlockDrop(
+				dropEventData,
+				targetRootClientId,
+				targetBlockIndex,
+				getBlockIndex,
+				getClientIdsOfDescendants,
+				moveBlocksToPosition,
+				insertBlocks,
+				clearSelectedBlock
+			);
+			setDropEventData( null );
+		}
+	}, [
+		dropEventData,
+		setDropEventData,
+		targetRootClientId,
+		targetBlockIndex,
+		getBlockIndex,
+		getClientIdsOfDescendants,
+		moveBlocksToPosition,
+		insertBlocks,
+		clearSelectedBlock,
+	] );
+
+	useEffect( () => {
+		if ( droppedFiles ) {
+			handleFilesDrop(
+				droppedFiles,
+				targetRootClientId,
+				targetBlockIndex,
+				hasUploadPermissions,
+				updateBlockAttributes,
+				canInsertBlockType,
+				insertBlocks
+			);
+			setDroppedFiles( null );
+		}
+	}, [
+		droppedFiles,
+		setDroppedFiles,
+		targetRootClientId,
+		targetBlockIndex,
+		hasUploadPermissions,
+		updateBlockAttributes,
+		canInsertBlockType,
+		insertBlocks,
+	] );
+
+	useEffect( () => {
+		if ( droppedHTML ) {
+			handleHTMLDrop(
+				droppedHTML,
+				targetRootClientId,
+				targetBlockIndex,
+				insertBlocks
+			);
+			setDroppedHTML( null );
+		}
+	}, [
+		droppedHTML,
+		setDroppedHTML,
+		targetRootClientId,
+		targetBlockIndex,
+		insertBlocks,
+	] );
+
+	// Optimization. Each callback returned from useOnBlockDrop should be a
+	// stable reference so that upstream in the useDropZone it isn't constantly
+	// unregistered and re-registered. These callbacks all primarily use the
+	// setter from `useState`, and delegate the handling of the drop event to
+	// an effect.
 	return {
-		onDrop: onBlockDrop(
-			targetRootClientId,
-			targetBlockIndex,
-			getBlockIndex,
-			getClientIdsOfDescendants,
-			moveBlocksToPosition,
-			insertBlocks,
-			clearSelectedBlock
-		),
-		onFilesDrop: onFilesDrop(
-			targetRootClientId,
-			targetBlockIndex,
-			hasUploadPermissions,
-			updateBlockAttributes,
-			canInsertBlockType,
-			insertBlocks
-		),
-		onHTMLDrop: onHTMLDrop(
-			targetRootClientId,
-			targetBlockIndex,
-			insertBlocks
+		onDrop: useCallback(
```

Comment:
If I'm understanding properly the purpose of the PR, I think we can avoid the useCallback entirely and avoid having the callback as a dependency. Instead we can just keep the latest value of the callback in a ref and call the ref in the callback.

We use this technique in several hooks already and I believe @ellatrix built a reusable hook somewhere to keep the ref "fresh" for us, right?

Reply:
We do this in several places, yes, but I don't think we have a utility function for it yet.